File: src/main/java/com/procurement/orchestrator/domain/dto/auction/ElectronicAuctionProgress.java
package com.procurement.orchestrator.domain.dto.auction;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Data;

import java.util.Set;

@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ElectronicAuctionProgress {

    @JsonProperty(value = "id")
    private String id;

    @JsonProperty(value = "period")
    private AuctionPeriod period;

    @JsonProperty(value = "breakdown")
    private Set<AuctionBreakdown> breakdown;

    public ElectronicAuctionProgress(@JsonProperty("id") final String id,
                                     @JsonProperty("period") final AuctionPeriod period,
                                     @JsonProperty("breakdown") final Set<AuctionBreakdown> breakdown) {
        this.id = id;
        this.period = period;
        this.breakdown = breakdown;
    }
}

File: src/main/kotlin/com/procurement/orchestrator/domain/model/tender/auction/ElectronicAuctionProgress.kt
package com.procurement.orchestrator.domain.model.tender.auction

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.procurement.orchestrator.domain.model.IdentifiableObject
import com.procurement.orchestrator.domain.model.period.Period
import com.procurement.orchestrator.domain.model.updateBy
import java.io.Serializable

data class ElectronicAuctionProgress(
    @field:JsonProperty("id") @param:JsonProperty("id") val id: ElectronicAuctionProgressId,

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    @field:JsonProperty("period") @param:JsonProperty("period") val period: Period? = null,

    @field:JsonInclude(JsonInclude.Include.NON_EMPTY)
    @field:JsonProperty("breakdown") @param:JsonProperty("breakdown") val breakdowns: ElectronicAuctionProgressBreakdowns = ElectronicAuctionProgressBreakdowns()
) : IdentifiableObject<ElectronicAuctionProgress>, Serializable {

    override fun equals(other: Any?): Boolean = if (this === other)
        true
    else
        other is ElectronicAuctionProgress
            && this.id == other.id

    override fun hashCode(): Int = id.hashCode()

    override fun updateBy(src: ElectronicAuctionProgress): ElectronicAuctionProgress {
        return ElectronicAuctionProgress(
            id = id,
            period = period updateBy src.period,
            breakdowns = breakdowns updateBy src.breakdowns
        )
    }
}
